spark-operator-api/src/main/java/org/apache/spark/k8s/operator/status/ApplicationStateSummary.java

```diff
@@ -0,0 +1,140 @@
+  /** Spark application is submitted to the cluster but yet scheduled */
+  Submitted,
+
+  /** Spark application will be restarted with same configuration */
+  ScheduledToRestart,
+
+  /** A request has been made to start driver pod in the cluster */
+  DriverRequested,
+
+  /**
+   * Driver pod has reached 'Running' state and thus bound to a node Refer
+   * https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/
+   */
+  DriverStarted,
+
+  /**
+   * Driver pod is ready to serve connections from executors Refer Refer
+   * https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/
+   */
+  DriverReady,
+
+  /**
+   * Less that minimal required executor pods reached condition 'Ready' during starting up Note that
+   * reaching 'Ready' does not necessarily mean that the executor has successfully registered with
+   * driver. This is a best-effort from operator to detect executor status
+   */
+  InitializedBelowThresholdExecutors,
+
+  /**
+   * All required executor pods started reached condition 'Ready' Note that reaching 'Ready' does
+   * not necessarily mean that the executor has successfully registered with driver. This is a
+   * best-effort from operator to detect executor status
+   */
+  RunningHealthy,
+
+  /** The application has lost a fraction of executors for external reasons */
+  RunningWithBelowThresholdExecutors,
+
+  /** The request timed out for driver */
+  DriverStartTimedOut,
+
+  /** The request timed out for executors */
+  ExecutorsStartTimedOut,
+
+  /** Timed out waiting for driver to become ready */
+  DriverReadyTimedOut,
+
+  /**
+   * The application completed successfully, or System.exit() is called explicitly with zero state
+   */
+  Succeeded,
+
+  /**
+   * The application has failed, JVM exited abnormally, or System.exit is called explicitly with
+   * non-zero state
+   */
+  Failed,
+
+  /**
+   * Operator failed to orchestrate Spark application in cluster. For example, the given pod
+   * template is rejected by API server because it's invalid or does not meet cluster security
+   * standard; operator is not able to schedule pods due to insufficient quota or missing RBAC
+   */
+  SchedulingFailure,
+
+  /** The driver pod was failed with Evicted reason */
+  DriverEvicted,
+
+  /** all resources (pods, services .etc have been cleaned up) */
+  ResourceReleased,
+
+  /**
+   * If configured, operator may mark app as terminated without releasing resources. While this can
+   * be helpful in dev phase, it shall not be enabled for prod use cases
+   */
+  TerminatedWithoutReleaseResources;
+
+  public boolean isInitializing() {
+    return Submitted.equals(this) || ScheduledToRestart.equals(this);
+  }
+
+  public boolean isStarting() {
+    return ScheduledToRestart.ordinal() < this.ordinal()
+        && RunningHealthy.ordinal() > this.ordinal();
+  }
+
+  public boolean isTerminated() {
+    return ResourceReleased.equals(this) || TerminatedWithoutReleaseResources.equals(this);
```

Comment:
I'm wondering if `isTerminated` considers the following.
- `ResourceRetainPolicy.Always`
- `ResourceRetainPolicy.OnFailure`